com.unity.ml-agents.extensions/Runtime/Teams/BaseTeamManager.cs

```diff
@@ -0,0 +1,166 @@
+using System;
+using System.Collections.Generic;
+
+namespace Unity.MLAgents.Extensions.Teams
+{
+    public class BaseTeamManager : ITeamManager, IDisposable
+    {
+        int m_StepCount;
+        int m_TeamMaxStep;
+        readonly int m_Id = TeamManagerIdCounter.GetTeamManagerId();
+        List<Agent> m_Agents = new List<Agent> { };
+
+
+        public BaseTeamManager()
+        {
+            Academy.Instance.TeamManagerStep += _ManagerStep;
+        }
+
+        public void Dispose()
+        {
+            Academy.Instance.TeamManagerStep -= _ManagerStep;
+        }
+
+        void _ManagerStep()
+        {
+            m_StepCount += 1;
+            if ((m_StepCount >= m_TeamMaxStep) && (m_TeamMaxStep > 0))
+            {
+                foreach (var agent in m_Agents)
+                {
+                    if (agent.enabled)
+                    {
+                        agent.EpisodeInterrupted();
+                    }
+                }
+                Reset();
+            }
+        }
+
+        /// <summary>
+        /// Register the agent to the TeamManager.
+        /// Registered agents will be able to receive team rewards from the TeamManager
+        /// and share observations during training.
+        /// </summary>
+        public virtual void RegisterAgent(Agent agent)
+        {
+            if (!m_Agents.Contains(agent))
+            {
+                m_Agents.Add(agent);
+            }
+        }
+
+        /// <summary>
+        /// Remove the agent from the TeamManager.
+        /// </summary>
+        public virtual void UnregisterAgent(Agent agent)
+        {
+            if (m_Agents.Contains(agent))
+            {
+                m_Agents.Remove(agent);
+            }
+        }
+
+        /// <summary>
+        /// Get the ID of the TeamManager.
+        /// </summary>
+        /// <returns>
+        /// TeamManager ID.
+        /// </returns>
+        public int GetId()
+        {
+            return m_Id;
+        }
+
+        /// <summary>
+        /// Get list of all agents registered to this TeamManager.
+        /// </summary>
+        /// <returns>
+        /// List of agents belongs to the TeamManager.
+        /// </returns>
+        public List<Agent> GetRegisteredAgents()
+        {
+            return m_Agents;
+        }
+
+        /// <summary>
+        /// Add team reward for all agents under this Teammanager.
+        /// Disabled agent will not receive this reward.
+        /// </summary>
+        public void AddTeamReward(float reward)
```

Comment:
I know that Agent currently has both AddReward and SetReward, but that causes some confusion among users. Can we just pick one and stick with it here?

Reply:
Had a discussion around this and we still think it's useful to have both and also keep the parity with agent Add/SetReward.